api/layer/layer.go

```diff
@@ -470,72 +471,76 @@ func getRandomOID() (*oid.ID, error) {
-	}
-	objVersion := &ObjectVersion{
-		BktInfo:               bkt,
-		ObjectName:            obj.Name,
-		VersionID:             obj.VersionID,
-		NoErrorOnDeleteMarker: true,
-	}
-
-	var nodeVersion *data.NodeVersion
-	nodeVersion, obj.Error = n.getNodeVersion(ctx, objVersion)
-
-	if obj.VersionID == UnversionedObjectVersionID {
-		if obj.Error == nil {
-			if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
-				return obj
-			}
-		} else if !errors.IsS3Error(obj.Error, errors.ErrNoSuchKey) {
+func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, settings *data.BucketSettings, obj *VersionedObject) *VersionedObject {
+	if len(obj.VersionID) != 0 {
+		var nodeVersion *data.NodeVersion
+		if nodeVersion, obj.Error = n.getNodeVersionToDelete(ctx, bkt, obj); obj.Error != nil {
 			return obj
 		}
 
-		randOID, err := getRandomOID()
-		if err != nil {
-			obj.Error = fmt.Errorf("couldn't get random oid: %w", err)
+		if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
 			return obj
 		}
 
-		newVersion := &data.NodeVersion{
-			BaseNodeVersion: data.BaseNodeVersion{
-				OID:      *randOID,
-				FilePath: obj.Name,
-			},
-			DeleteMarker: &data.DeleteMarkerInfo{
-				Created: time.Now(),
-				Owner:   n.Owner(ctx),
-			},
-			IsUnversioned: true,
-		}
+		obj.Error = n.treeService.RemoveVersion(ctx, &bkt.CID, nodeVersion.ID)
+		return obj
+	}
 
-		if obj.Error = n.treeService.AddVersion(ctx, &bkt.CID, newVersion); obj.Error != nil {
-			return obj
-		}
+	var newVersion *data.NodeVersion
 
-		n.namesCache.Delete(bkt.Name + "/" + obj.Name)
-	} else {
-		if obj.Error != nil {
-			return obj
-		}
+	if !settings.VersioningEnabled {
+		obj.VersionID = UnversionedObjectVersionID
 
-		if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
+		var nodeVersion *data.NodeVersion
+		if nodeVersion, obj.Error = n.getNodeVersionToDelete(ctx, bkt, obj); obj.Error != nil {
 			return obj
 		}
 
-		if obj.Error = n.treeService.RemoveVersion(ctx, &bkt.CID, nodeVersion.ID); obj.Error != nil {
+		if obj.Error == nil {
+			if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
+				return obj
+			}
+		} else if !errors.IsS3Error(obj.Error, errors.ErrNoSuchKey) {
 			return obj
 		}
 	}
 
-	n.listsCache.CleanCacheEntriesContainingObject(obj.Name, bkt.CID)
+	randOID, err := getRandomOID()
+	if err != nil {
+		obj.Error = fmt.Errorf("couldn't get random oid: %w", err)
+		return obj
+	}
+	newVersion = &data.NodeVersion{
+		BaseNodeVersion: data.BaseNodeVersion{
+			OID:      *randOID,
+			FilePath: obj.Name,
+		},
+		DeleteMarker: &data.DeleteMarkerInfo{
+			Created: time.Now(),
+			Owner:   n.Owner(ctx),
+		},
+		IsUnversioned: !settings.VersioningEnabled,
+	}
+
+	if obj.Error = n.treeService.AddVersion(ctx, &bkt.CID, newVersion); obj.Error != nil {
```

Comment:
Why we need to hang delete marker on a tree when bucket is unversioned? 
If I understand correctly we should create delete marker for versioned object. https://docs.aws.amazon.com/AmazonS3/latest/userguide/DeleteMarker.html

Reply:
AWS do the same:
```
$ aws s3api --profile aws get-bucket-versioning --bucket dkirillov 
{
    "Status": "Suspended"
}
$ aws s3api --profile aws put-object --bucket dkirillov --key a
{
    "ETag": "\"d41d8cd98f00b204e9800998ecf8427e\""
}
$ aws s3api --profile aws list-object-versions --bucket dkirillov 
{
    "Versions": [
        {
            "ETag": "\"d41d8cd98f00b204e9800998ecf8427e\"",
            "Size": 0,
            "StorageClass": "STANDARD",
            "Key": "a",
            "VersionId": "null",
            "IsLatest": true,
            "LastModified": "2022-06-30T06:35:52+00:00",
            "Owner": {
                "ID": "68e12342e5971d09ced74634fc7cc0fc5f01419d06b76353d42ce3108367183b"
            }
        }
    ]
}
$ aws s3api --profile aws delete-object --bucket dkirillov --key a
{
    "DeleteMarker": true,
    "VersionId": "null"
}
$ aws s3api --profile aws list-object-versions --bucket dkirillov 
{
    "DeleteMarkers": [
        {
            "Owner": {
                "ID": "68e12342e5971d09ced74634fc7cc0fc5f01419d06b76353d42ce3108367183b"
            },
            "Key": "a",
            "VersionId": "null",
            "IsLatest": true,
            "LastModified": "2022-06-30T06:36:22+00:00"
        }
    ]
}
```

Reply:
Now there is no difference between `suspended` and `unversioned` mode. It should be reconsidered because it seems like we can differ these modes with the new tree service.